### ISSUE TITLE
[Train at Home 3.7.0] Repeated NaN loss followed by 100% NaN pseudo-gradients and miner shutdown on M2 Pro

### ISSUE THREAD

**Issue by shinteni** [macrocosm-os/iota#198]
## Problem

Train at Home 3.7.0 on a MacBook (Apple M2 Pro, 32 GB RAM, macOS 26.5.2) has twice encountered NaN loss, continued skipping affected batches, and then exited when all 506,021,888 pseudo-gradient values failed the submission-time NaN check. A fresh training session subsequently failed with the same symptom.

The live trigger has **not** been reproduced. The evidence below distinguishes observed failures from an offline reproduction of a NaN propagation path. This report was prepared with AI-assisted log/source analysis; the propagation experiment was isolated from the running miner.

## Observed incidents

All times are JST (UTC+09:00).

**2026-09-12, run `4.12.16.2-tah`, layer 2:**

- 13:11:15.703: backward completed with finite loss `6.4204583168029785`.
- 13:11:15.796–13:11:16.226: gradient clipping and optimizer update completed; log reports `Optimization step #100`.
- 13:11:25.107: the next batch reported NaN loss.
- Eight batches were subsequently skipped with `Loss ... has NaNs`; no further `BACKWARD complete` appeared before shutdown.
- 13:15:55–13:15:56: previous weights had finite mean/std (`-0.000931`, `0.035156`); new weights had NaN mean/std.
- 13:16:03.650: `506021888 / 506021888 = 100.00%` pseudo-gradients were NaN.
- 13:19:48.296: `Miner shutdown complete`.

**2026-09-13, run `4.12.16.5-tah`, layer 2:**

- This session had successful weight submissions at 2026-09-12 22:03:41, 2026-09-13 00:51:51 and 03:46:38.
- NaN loss and skipped batches were again observed, including at 14:23:33 and 14:24:13.
- 14:26:05.646: the same `506021888 / 506021888 = 100.00%` pseudo-gradient NaN check failed.
- 14:26:05.724: `Critical error in run_miner: pseudo gradients ... has NaNs`.
- 14:28:03.124: `Miner shutdown complete`. Subsequent process inspection found no training worker.

## Source observations and scope

The installed bytecode for six relevant methods matches public commit `22d702acd07453dabb187abedfad185826091341`: `_backward`, `clip_gradients`, `local_optimization_step`, `compute_loss`, `submit_weights`, and `_run_backward_slices`.

- [`clip_gradients`](https://github.com/macrocosm-os/iota/blob/22d702acd07453dabb187abedfad185826091341/shared/subnet/src/subnet/model/model_mixin.py#L294-L302) calls `clip_grad_norm_` without `error_if_nonfinite=True`. The update path then calls `optimizer.step()` without checking parameter gradients for finiteness. A finite scalar loss does not establish finite gradients after backward.
- [The loss-error handler](https://github.com/macrocosm-os/iota/blob/22d702acd07453dabb187abedfad185826091341/src/miner/src/miner/training/training.py#L493-L515) catches the NaN loss exception and skips the batch, without restoring the model on that path.
- Full numerical scans of the first incident's saved 10:59:25 baseline weights and optimizer state found zero NaNs/Infs; second moments were nonnegative. These snapshots predate the failure and do **not** establish that the gradients entering step 100 were finite.

The first invalid value's originating operation is unknown: the relevant live gradients and before/after-update parameter snapshots were not captured. This is not a claim that gradient clipping was conclusively the live trigger or that a particular MPS kernel is faulty. It also does not establish a causal connection to #196.

## Offline propagation reproduction

Tested on CPU with BF16 tensors and PyTorch 2.10.0, git `449b1768410104d3ed79d3bcfe4ba1d65c7f22c0` (matching the installed worker's PyTorch version). The isolated experiment used the extracted matching `clip_gradients` method and the observed AdamW settings. Equivalent standalone example:

```python
import torch

for inject_nan in (False, True):
    params = [torch.nn.Parameter(torch.ones(4, dtype=torch.bfloat16))
              for _ in range(2)]
    opt = torch.optim.AdamW(params, lr=0.0002, betas=(0.9, 0.95),
                            eps=1e-8, weight_decay=0.1)
    for p in params:
        p.grad = torch.ones_like(p)
    if inject_nan:
        params[0].grad[0] = float('nan')
    torch.nn.utils.clip_grad_norm_(params, max_norm=1.0)
    bad_grads = sum(torch.isnan(p.grad).sum().item() for p in params)
    opt.step()
    bad_weights = sum(torch.isnan(p).sum().item() for p in params)
    print(inject_nan, bad_grads, bad_weights)
```

Observed output: `False 0 0`, then `True 8 8`, with no exception. One injected NaN contaminated all eight test gradients and weights. This reproduces propagation only, not the first NaN's production origin or a full MPS training failure.

## Requested maintainer guidance

Could you investigate the first invalid-gradient/update source and provide a supported fix or diagnostic build? In particular, early nonfinite detection around backward/clipping/optimizer updates and a defined recovery policy would help avoid repeatedly using a corrupted model until submission-time shutdown. Please advise which sanitized diagnostics would identify the original operation without changing registration, attestation, or training protocol behavior.

No wallet secrets, authentication material, raw device logs, or production model files are attached. No application patch or live fault injection was performed for this investigation.

**Comment by peterhon168** [comment]
Follow-up: another live occurrence on 2026-09-17 (JST)

This is another occurrence of the same failure on the same M2 Pro / 32 GB Mac, now on run `4.12.16.4-tah`, layer 2. No raw logs, wallet data, authentication material, or production model files are attached.

### Observed timeline

- At 19:46:51 the client saved a current model/optimizer snapshot. At submission time, the previous weights from this finite snapshot still had mean `-0.000343` and std `0.092773`.
- At 19:53:11, `Optimization step #4` completed. The two preceding scalar losses were finite (`5.461079` and `5.808987`).
- At 19:54:10, the next last-layer loss was NaN for activation `6dd35fca...`; the logged output shape was `[4, 800, 128256]` and target shape `[4, 800]`.
- From 19:54:10 through 20:27:33, at least 18 different activations reported `Loss ... has NaNs`. The handler skipped backward for each one, but no model/optimizer rollback occurred.
- At 20:28:02, pseudo-gradient mean and std were both `nan`; at 20:28:04, the previous weights were still finite but the new weights had mean/std `nan`.
- At 20:29:33, the submission-time check found `506021888 / 506021888 = 100.00%` NaN pseudo-gradient values. Weight submission failed with `NanInfException`, followed by an orderly miner shutdown.
- The process exited after cleanup with code 0, but the next start saw `Entity not registered` and obtained a new registration queue entry. The previous active assignment was not resumed.

### Additional implementation observation

The installed bytecode for `ModelManager.clip_gradients` calls `torch.nn.utils.clip_grad_norm_(parameters, max_norm)` without passing `error_if_nonfinite=True` (matching the public source at [model_mixin.py#L294-L302](https://github.com/macrocosm-os/iota/blob/22d702acd07453dabb187abedfad185826091341/shared/subnet/src/subnet/model/model_mixin.py#L294-L302)). In the bundled PyTorch implementation, the default is `error_if_nonfinite=False`.

This means that if one accumulated gradient is non-finite, the global norm and clipping coefficient can become non-finite; the in-place clipping can then contaminate all accumulated gradients before `optimizer.step()`. This explains how a local invalid gradient can become the observed 100% NaN weights/pseudo-gradients. It does not prove which activation, target, logit, backward operation, or MPS kernel produced the first invalid value.

Earlier in the same run, global optimizer-state download also failed with a non-fatal warning and training continued with the locally initialized optimizer. Causality is not established, but please clarify whether this fallback is expected to be numerically safe.

### Requested fix / diagnostic guidance

Could you provide a supported fix or diagnostic build that:

1. checks activation/target/logit/loss finiteness before backward;
2. rejects a non-finite total gradient norm before clipping;
3. clears the entire accumulation window when a bad gradient is detected;
4. validates parameters and optimizer state after `optimizer.step()` and restores the last finite checkpoint on failure; and
5. records sanitized dtype/device/shape and finite-count diagnostics sufficient to identify the first invalid tensor?

The goal is to avoid continuing with a corrupted model until submission-time shutdown and to avoid losing the active registration assignment. No registration, attestation, or training-protocol bypass was attempted.
